### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by 3zcurdia**
### The purpose of following comments and modifications is to provide a positive feedback to the author, therefore he could become a better developer.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,77 +1,36 @@
 source 'https://rubygems.org'
 
-# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
+ruby '2.2.3'
```
Comment (3zcurdia) [inline Gemfile]: Is nice to have a ruby version in the gemfile, services as heroku took the ruby version from this file

File: Gemfile [hunk 2]
```diff
@@ -1,77 +1,36 @@
...
+
 gem 'rails', '4.2.5'
+gem 'pg_search'
+gem 'pg', '~> 0.15'
```
Comment (3zcurdia) [inline Gemfile]: Is nice to have database gems next to the rails gem

File: Gemfile [hunk 3]
```diff
@@ -1,77 +1,36 @@
...
 gem 'coffee-rails', '~> 4.1.0'
-# See https://github.com/rails/execjs#readme for more supported runtimes
-# gem 'therubyracer', platforms: :ruby
-
-# Turbolinks makes following links in your web application faster. Read more: https://github.com/rails/turbolinks
 gem 'turbolinks'
-# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
 gem 'jbuilder', '~> 2.0'
-# bundle exec rake doc:rails generates the API under doc/api.
-gem 'sdoc', '~> 0.4.0', group: :doc
```
Comment (3zcurdia) [inline Gemfile]: this gem is never used

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -8,13 +8,10 @@ class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
 
   def default_url_options(options = {})
-    {locale: set_locale}.merge options
+    { locale: set_locale }.merge options
```
Comment (3zcurdia) [inline app/controllers/application_controller.rb]: always leave spaces

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -1,12 +1,12 @@
 class CommentsController < ApplicationController
-  before_filter :signed_user_only, :only => [:create]
+  before_filter :signed_user_only, only: [:create]
```
Comment (3zcurdia) [inline app/controllers/comments_controller.rb]: use this symbol syntax for options and hashes

File: app/controllers/comments_controller.rb [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
 
   def create
     post = Post.friendly.find(params[:post_id])
     comment = post.comments.build(comments_params)
     comment.user_id = current_user.id
     if !comment.save
-
+      # Do not leave empty blocks
```
Comment (3zcurdia) [inline app/controllers/comments_controller.rb]: do not  leave empty blocks and always use positive logic is easy to read

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -47,13 +46,14 @@ def destroy
...
+
   # Use callbacks to share common setup or constraints between actions.
   def set_post
-    begin
-      @post = Post.friendly.find(params[:id])
-    rescue ActiveRecord::RecordNotFound
-      redirect_to :root
-    end
+    @post = Post.friendly.find(params[:id])
+  rescue ActiveRecord::RecordNotFound
```
Comment (3zcurdia) [inline app/controllers/posts_controller.rb]: if you are rescuing the whole method you can use this syntax

File: app/form_builders/bootstrap/form_builder_normal.rb [hunk 8]
```diff
@@ -39,25 +40,26 @@ def radio_button(tag_value, options = {})
...
         super(method, collection, value_method, text_method, options, html_options)
       end
     end
 
     private
+
     def add_class_form_control!(options)
       if options[:class]
-        options[:class] = 'form-control '+options[:class]
+        options[:class] = 'form-control ' + options[:class]
```
Comment (3zcurdia) [inline app/form_builders/bootstrap/form_builder_normal.rb]: for string interpolation is better to use 

    "form-control #{options[:class]}

it performs better

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -1,4 +1,8 @@
 module SessionsHelper
+  # Check the following gems for authentication:
+  # * devise
```
Comment (3zcurdia) [inline app/helpers/sessions_helper.rb]: https://github.com/plataformatec/devise

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -1,4 +1,8 @@
...
+  # * warden
```
Comment (3zcurdia) [inline app/helpers/sessions_helper.rb]: https://github.com/hassox/warden

File: app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -1,4 +1,8 @@
...
+  # * cancancan (authorization)
```
Comment (3zcurdia) [inline app/helpers/sessions_helper.rb]: https://github.com/CanCanCommunity/cancancan

File: app/models/category.rb [hunk 12]
```diff
@@ -1,3 +1,4 @@
 class Category < ActiveRecord::Base
+  # For categories trees check act_as_tree
```
Comment (3zcurdia) [inline app/models/category.rb]: https://github.com/amerine/acts_as_tree

File: app/models/post.rb [hunk 13]
```diff
@@ -5,13 +5,16 @@ class Post < ActiveRecord::Base
   include PgSearch
   extend FriendlyId
 
-  before_save :config_initialization
+  after_initialize :config_initialization
```
Comment (3zcurdia) [inline app/models/post.rb]: If you are initializing something you should use this callback

File: app/models/post.rb [hunk 14]
```diff
@@ -21,17 +24,15 @@ def self.full_text_search(query)
...
+
   def locale_to_config
-    case I18n.locale
-      when :es
-        'spanish'
-      else
-        'english'
-    end
+    'spanish' if I18n.locale == :es
+    'english'
```
Comment (3zcurdia) [inline app/models/post.rb]: switch is not needed

File: app/models/user.rb [hunk 15]
```diff
@@ -1,13 +1,14 @@
...
   has_secure_password
-
   has_many :posts
 
-  validates :username, :password, :password_confirmation, :email, :presence => true
-  validates :username, :email, :uniqueness =>  true
+  validates :username, :password, :password_confirmation, :email, presence: true
+  validates :username, :email, uniqueness:  true
+
+  before_save :donwcase_email_username
```
Comment (3zcurdia) [inline app/models/user.rb]: Is more maintainable have methods in callbacks than blocks

File: bin/setup [hunk 16]
```diff
@@ -2,15 +2,15 @@
 require 'pathname'
```
Comment (3zcurdia) [inline bin/setup]: the corrections here were product of a linter

File: config/database.yml [hunk 17]
```diff
@@ -29,10 +29,8 @@ development:
   # To create additional roles in postgres see `$ createuser --help`.
   # When left blank, postgres will use the default role. This is
   # the same name as the operating system user that initialized the database.
-  username: mblog
 
   # The password associated with the postgres role (username).
-  password: m54ncjF!
```
Comment (3zcurdia) [inline config/database.yml]: never leave passwords on .yml files

File: db/migrate/20160117003902_add_config_to_post_translations.rb [hunk 18]
```diff
@@ -1,11 +1,9 @@
 class AddConfigToPostTranslations < ActiveRecord::Migration
   def up
-    execute "ALTER TABLE post_translations ADD COLUMN config regconfig"
+    execute 'ALTER TABLE post_translations ADD COLUMN config regconfig'
```
Comment (3zcurdia) [inline db/migrate/20160117003902_add_config_to_post_translations.rb]: use rails callbacks

    add_column :post_transactions, :config

File: rails_best_practices_output.html [hunk 19]
```diff
@@ -0,0 +1,320 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
```
Comment (3zcurdia) [inline rails_best_practices_output.html]: This file was renerated by rails best practices gem, open this file on the browser

File: app/views/layouts/_foot.html.erb [hunk 20]
```diff
@@ -1,4 +1,4 @@
 <br />
-<%= debug(params) if Rails.env.development? %>
+<%#= debug(params) if Rails.env.development? %>
```
Comment (3zcurdia) [inline app/views/layouts/_foot.html.erb]: If you want to debug the params you could do it by using debug on the controller, leaving debuggers on commits is a bad code smell

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -1,4 +1,6 @@
 class UsersController < ApplicationController
+  # Security Hole!!!
+  # On this controller you can edit users without any session active
```
Comment (3zcurdia) [inline app/controllers/users_controller.rb]: # Security hole!!!
On this controller you can edit users without any session active

File: app/form_builders/bootstrap/form_builder_normal.rb [hunk 22]
```diff
@@ -1,34 +1,35 @@
 module Bootstrap
+  # this whole formbuilder could be avoided using simple form
```
Comment (3zcurdia) [inline app/form_builders/bootstrap/form_builder_normal.rb]: https://github.com/plataformatec/simple_form

File: app/models/post.rb [hunk 23]
```diff
@@ -5,13 +5,16 @@ class Post < ActiveRecord::Base
...
   translates :title, :body, :config, :slug
-  friendly_id :title, :use => :globalize
+  friendly_id :title, use: :globalize
 
   mount_uploader :image, PostImageUploader
 
   def self.full_text_search(query)
+    # This is a Clever solution full text search is very complex to implement
+    # it will be more easy to scale if you are using a tool like pg_search
+    # just be mindfull that all code that you make is going to be extended by other devs
```
Comment (3zcurdia) [inline app/models/post.rb]: :+1:  for the solution 
:-1:  for the scalability

File: app/models/post.rb [hunk 24]
```diff
@@ -21,17 +24,15 @@ def self.full_text_search(query)
...
+
   def locale_to_config
-    case I18n.locale
-      when :es
-        'spanish'
-      else
-        'english'
-    end
+    return 'spanish' if I18n.locale == :es
+    'english'
```
Comment (3zcurdia) [inline app/models/post.rb]: switch is not needed here

File: app/models/post.rb [hunk 25]
```diff
@@ -21,17 +24,15 @@ def self.full_text_search(query)
...
   end
 
   def config_initialization
-    self.config = locale_to_config
+    # Aviable languages are alredy defined on application.rb
```
Comment (3zcurdia) [inline app/models/post.rb]: Aviable languages are alredy defined on application.rb

File: app/views/layouts/application.html.erb [hunk 26]
```diff
@@ -1,3 +1,4 @@
+<!-- Application layout should have header and footer -->
```
Comment (3zcurdia) [inline app/views/layouts/application.html.erb]: Application layout should be in a single file, avoid unnecessary view renders

File: app/views/layouts/posts.html.erb [hunk 27]
```diff
@@ -1,3 +1,4 @@
+<!-- This view should be under post controller -->
```
Comment (3zcurdia) [inline app/views/layouts/posts.html.erb]: This is not a layout, this sould be a view inside the post views directory

File: app/models/post.rb (hunk 23 above)
Comment (sounxrt) [inline app/models/post.rb]: I'm not sure if using pg_search + globalize will use the indexes correctly according to the documentation of pg_search but i need to research more, thanks anyway

**Comment by sounxrt** [comment]
Thank you very much for your comments and suggestions

Regards

**Pull request merged**